### ISSUE TITLE
Timing issue ? Timeout errors... [SOLVED]

### ISSUE THREAD

**Issue by yet-another-average-joe** [4-20ma/ModbusMaster#126]
Hello,

I had a problem with AVRs (Nano, Pro Mini, Mega), Cortex (Due) and STM32 (Blue Pill, Black Pill) : timeout errors (226) if modbus function calls are too near from each other.

See also : https://github.com/4-20ma/ModbusMaster/issues/108

Not having a logic analyzer when I first used this library (just a scope, 2 years ago), I had a hard time ! The Samae is now on the way ! Without a logic analyzer, I still don't really know why it works.

Tinkering with delays, I found 2 "dirty fixes", that worked for me :

```

void preTransmission()
{
	delay(3);
	digitalWrite(MAX485_RE, HIGH);
#if MAX485_RE != MAX485_DE
	digitalWrite(MAX485_DE, HIGH);
#endif
}
```

OR (I first did this :)

`delay(5);
`

after each and every Modbus function call. For my projects, I need the calls to be as near as possible from each others, and I'm limited @ 19200 bps by a VFD. (commands must be responsive (better than 100ms, with 4 holding registers, that will be reduced to 2, playing with some bit packing).

The problem does not depend on the slave. It is the same with a Schneider variable frequency drive, and Arduino slaves running SimpleModbusSlave.

The problem does not depend on the platform (all MC I tested gave the same errors)

With the VFD, the project reads/writes 55 holding registers in a loop, as fast as possible...
With the other slaves, it reads /writes a couple buttons and coils
19200 bauds

This being said, this library is really great : easy to use and self explanative ! (I tested all Arduino oriented libraries : it is the best, IMHO, and by far). And I just realize I forgot to star 2 years ago ! Done.

**Comment by yet-another-average-joe** [comment]
After some reading about Modbus RTU specification...

There should be at least 3.5 char (t3.5) silent interval between two frames. This delay is not implemented  in ModbusMaster.

Modbus RTU specification and calculations :

1 start bit + 8 data bits + 1 parity bit + 1 stop bit
or
1 start bit + 8 data bits + (no parity bit) + 2 stop bits

total : 11 bits

above 19200 bps : t3.5 must be fixed @ 1750 us
up to 19200 bps (included) : 3.5 * 11 = 38.5 bits -> (39 * bit_duration)
bit duration = 1 / baud_rate

-> **delay to be added = t3.5 (us) = 1,000,000 * 39 / baud_rate**

```
#if MODBUS_BAUD_RATE > 19200
        uint32_t t3_5 = 1750;
#else
        uint32_t t3_5 = (1000000 * 39) / MODBUS_BAUD_RATE + 500; // +500us : to be safe
#endif

void preTx()
{
	// silent interval between 2 frames
	// 1750us (> 19200bps) or t3.5 (<= 19200pbs)
	delayMicroseconds(t3_5);

	digitalWrite(MAX485_RE, HIGH);

        // generally, RE and DE are connected together on the MAX485...
#if MAX485_RE != MAX485_DE
	digitalWrite(MAX485_DE, HIGH);
#endif
}
```
The delayMicroseconds  is called in pre transmission callback. I couldn't get it to work in the post transmission callback. I didn't investigate. Anyway, the result is the same : frame will be separated by the required delay.

Obviously, this is not needed if the software computes for more than this duration between Modbus transmissions...

**Comment by yet-another-average-joe** [comment]
Closed.

See : https://github.com/4-20ma/ModbusMaster/issues/127
